src/envelope.rs

```diff
@@ -81,48 +125,99 @@ impl<D: Hash> Envelope<D> {
     /// nonce             | ciphertext       | hmac
     /// nonce_size bytes  | variable length  | hmac_size bytes
     pub(crate) fn from_bytes(bytes: &[u8]) -> Result<Self, InternalPakeError> {
-        let ciphertext_start = NONCE_LEN;
-        let ciphertext_end = bytes.len() - Self::hmac_size();
-
-        Ok(Self::new(
-            bytes[..ciphertext_start].to_vec(),
-            bytes[ciphertext_start..ciphertext_end].to_vec(),
-            GenericArray::clone_from_slice(&bytes[ciphertext_end..]),
-        ))
+        let (result, remainder) = Self::deserialize(bytes)
+            .map_err(|_| InternalPakeError::IncompatibleEnvelopeCredentialsError)?;
+        if !remainder.is_empty() {
+            return Err(InternalPakeError::IncompatibleEnvelopeCredentialsError);
+        }
+        Ok(result)
     }
 
     pub(crate) fn to_bytes(&self) -> Vec<u8> {
-        [&self.nonce[..], &self.ciphertext[..], &self.hmac[..]].concat()
+        self.serialize()
     }
 
     pub(crate) fn serialize(&self) -> Vec<u8> {
         [
             &self.nonce[..],
             &serialize((&self.ciphertext).to_vec(), 2)[..],
-            &serialize(vec![], 2)[..],
+            &serialize((&self.auth_data).to_vec(), 2)[..],
             &serialize((&self.hmac).to_vec(), 2)[..],
         ]
         .concat()
     }
 
     pub(crate) fn deserialize(input: &[u8]) -> Result<(Self, Vec<u8>), ProtocolError> {
+        if input.len() < NONCE_LEN {
+            return Err(ProtocolError::VerificationError(
+                PakeError::SerializationError,
+            ));
+        }
+
         let nonce = &input[..NONCE_LEN];
         let (ciphertext, remainder) = tokenize(input[NONCE_LEN..].to_vec(), 2)?;
-        let (_, remainder) = tokenize(remainder, 2)?;
+        let (auth_data, remainder) = tokenize(remainder, 2)?;
         let (hmac, remainder) = tokenize(remainder, 2)?;
         Ok((
             Self::new(
                 nonce.to_vec(),
                 ciphertext,
+                auth_data,
                 GenericArray::clone_from_slice(&hmac[..]),
             ),
             remainder,
         ))
     }
 
+    fn serialize_extensions(
+        cred_format: Vec<CredentialType>,
+        credentials: &HashMap<CredentialType, Vec<u8>>,
+    ) -> Result<Vec<u8>, InternalPakeError> {
+        let mut ret = Vec::new();
+        for index_type in cred_format {
+            let byte: u8 = index_type as u8 + 1;
+            let chunk = match &credentials.get(&index_type) {
+                Some(v) => [vec![byte], serialize(v.to_vec(), 2)].concat(),
+                None => return Err(InternalPakeError::IncompatibleEnvelopeCredentialsError),
+            };
+            ret.extend(chunk);
+        }
+        Ok(ret)
```

Comment:
```suggestion
        let mut ret = Vec::new();
        for index_type in cred_format {
            match &credentials.get(&index_type) {
                Some(v) => {
                    ret.push(index_type as u8 + 1);
                    ret.extend(serialize(v.to_vec(), 2));
                }
                None => return Err(InternalPakeError::IncompatibleEnvelopeCredentialsError),
            }
        }
        Ok(ret)
```

Also `serialization::serialize` should clearly be fixed to take its argument by reference.